clients/drcachesim/tests/schedule_stats_test.cpp

```diff
@@ -0,0 +1,208 @@
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, LLC OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Test for checks performed by invariant_checker_t that are not tested
+ * by the signal_invariants app's prefetch and handler markers.
+ * This looks for precise error strings from invariant_checker.cpp: but
+ * we will notice if the literals get out of sync as the test will fail.
+ */
+
+#undef NDEBUG
+#include <assert.h>
+
+#include <fstream>
+#include <iostream>
+#include <unordered_map>
+#include <vector>
+
+#include "../tools/schedule_stats.h"
+#include "../common/memref.h"
+#include "memref_gen.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+using ::dynamorio::drmemtrace::default_memtrace_stream_t;
+using ::dynamorio::drmemtrace::memref_t;
+using ::dynamorio::drmemtrace::memref_tid_t;
+using ::dynamorio::drmemtrace::TRACE_MARKER_TYPE_CORE_WAIT;
+using ::dynamorio::drmemtrace::TRACE_MARKER_TYPE_DIRECT_THREAD_SWITCH;
+using ::dynamorio::drmemtrace::TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL;
+using ::dynamorio::drmemtrace::TRACE_MARKER_TYPE_SYSCALL;
+
+// Bypasses the analyzer and scheduler for a controlled test sequence.
+// Alternates the per-core memref vectors in lockstep.
+static schedule_stats_t::counters_t
+run_schedule_stats(const std::vector<std::vector<memref_t>> &memrefs,
+                   const std::unordered_map<memref_tid_t, int64_t> &tid2ord)
+{
+    schedule_stats_t tool(/*print_every=*/1, /*verbosity=*/2);
+    // schedule_stats_t uses get_input_id() to identify switches.
+    class mock_stream_t : public default_memtrace_stream_t {
+    public:
+        void
+        set_input_id(int64_t input_id)
+        {
+            input_id_ = input_id;
+        }
+        int64_t
+        get_input_id() const override
+        {
+            return input_id_;
+        }
```

Comment:
We can potentially override get_input_interface here to return another default_memtrace_stream_t, which will obviate the need for the test-specific check `shard->stream->get_input_interface() != nullptr`.

Reply:
Done.